nengo/transforms.py

```diff
@@ -0,0 +1,348 @@
+import numpy as np
+
+import nengo
+from nengo.base import FrozenObject
+from nengo.dists import Distribution, DistOrArrayParam
+from nengo.exceptions import ValidationError
+from nengo.params import Parameter, ShapeParam, IntParam, EnumParam, BoolParam
+from nengo.utils.compat import is_array_like
+
+
+class Transform(FrozenObject):
+    """A base class for connection transforms."""
+
+    def sample(self, rng=np.random):
+        """
+        Returns concrete weights needed to implement the specific transform.
+
+        Parameters
+        ----------
+        rng : `numpy.random.RandomState`, optional
+            Random number generator state.
+
+        Returns
+        -------
+        array_like
+            Transform weights
+        """
+        raise NotImplementedError()
+
+    @property
+    def size_in(self):
+        """Expected size of input to transform"""
+        raise NotImplementedError()
+
+    @property
+    def size_out(self):
+        """Expected size of output from transform"""
+        raise NotImplementedError()
+
+
+class TransformParam(Parameter):
+    """A parameter where the value must be a Transform.
+
+    Also supports casting array_likes to Dense transforms.
+    """
+
+    coerce_defaults = False
+
+    def __init__(self, name, default, optional=False, readonly=False):
+        super(TransformParam, self).__init__(
+            name, default, optional, readonly)
+
+    def coerce(self, conn, transform):
+        if not isinstance(transform, Transform):
+            transform = Dense((conn.size_out, conn.size_mid), transform)
```

Comment:
Rather than having this be tied to a Connection (in which case, it really seems like something that should live in `connection.py`), could we have a `shape` param where we pass `('size_out', 'size_mid')` like with `NdArrayParam.shape` for Connections? Or maybe we'd want them as separate `size_in` and `size_out` params. Then, this could potentially be used other places.

One potential place it could immediately be used is for `Ensemble.encoders`. It would be nice (for me at least) to be able to have convolutional encoders. This might be a bit of work, so I'm not saying it should happen in this PR necessarily. And maybe the suggestion I made about generalizing TransformParam shouldn't be implemented until we're using it in another place so that we really have a general test for it. But long term, that's where I see all of this going.

Maybe I'll take a quick look at that now and see how much work it would be.